File: bubble_sort/bubbleSort.cpp
/**
 * bubble sort
 * @Author alacic gu
 */

#include <iostream>

using namespace std;

void swap(int *a, int *b);
void bubbleSort(int arr[], int arr_size);
void printArray(int arr[], int arr_size);

int main()
{
    int arr[]{1, 5, 9, 7, 8, 12, 5, 5, 5, 2};
    cout << "Before sorted, arr is ";
    int n = sizeof(arr) / sizeof(arr[0]);
    printArray(arr, n);
    bubbleSort(arr, n);
    cout << "After sorted, arr is ";
    printArray(arr, n);
}

void swap(int *a, int *b)
{
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

void bubbleSort(int arr[], int arr_size)
{
    for (int i = 0; i < arr_size; i++)
    {
        for (int j = 0; j < arr_size - i - 1; j++)
        {
            if (arr[j] > arr[j + 1])
            {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

void printArray(int arr[], int arr_size)
{
    for (int i = 0; i < arr_size - 1; i++)
    {
        cout << arr[i] << ", ";
    }
    cout << arr[arr_size - 1] << endl;
}

File: bubble_sort/bubble_sort.py
# @Author alacic gu


def bubble_sort(arr):
    arr_size = len(arr)
    for i in range(arr_size):
        for j in range(arr_size-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]


if __name__ == "__main__":
    arr = [1, 5, 6, 3, 4, 7, 8, 9, 5, 4, 7, 5]
    print(f"Before sorted, arr is {arr}")
    bubble_sort(arr)
    print(f"After sorted, arr is {arr}")
